flowmachine/flowmachine/features/subscriber/home_location_monthly.py

```diff
@@ -0,0 +1,198 @@
+        return self._window_start.strftime("%Y-%m-%d")
+
+    @window_start.setter
+    def window_start(self, value):
+        if type(value) is str:
+            self._window_start = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) is datetime:
+            self._window_start = value
+        else:
+            raise TypeError("window_start must be datetime or yyyy-mm-dd")
+
+    @property
+    def window_stop(self):
+        return self._window_stop.strftime("%Y-%m-%d")
+
+    @window_stop.setter
+    def window_stop(self, value):
+        if type(value) is str:
+            self._window_stop = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) is datetime:
+            self._window_stop = value
+        else:
+            raise TypeError("window_stop must be datetime or yyyy-mm-dd")
+
+    def _make_query(self):
+
+        last_locations_clause = ""
+
+        # NOTE: last_locations seems to be dropping the final record? This is weird.
+        for last_location in self.last_locations:
+            last_locations_clause += f"""
+SELECT subscriber, pcod, '{last_location.start}' AS day
+FROM ({last_location.get_query()}) AS tbl
+UNION ALL
+"""
+
+        last_locations_clause = last_locations_clause.rstrip("UNION ALL\n")
+
+        sql = f"""
+WITH last_locations AS (
+    {last_locations_clause}
+), modal_locations AS (
+    {self.modal_locations.get_query()}
+), location_histogram AS (
+	SELECT subscriber, pcod, count(pcod) as y
+	FROM last_locations
+	GROUP BY subscriber, pcod
+), this_month_known_homes AS (
+	SELECT modal_locations.subscriber, modal_locations.pcod 
+	FROM modal_locations
+		INNER JOIN  location_histogram ON location_histogram.subscriber = modal_locations.subscriber
+	WHERE modal_locations.pcod = location_histogram.pcod
+	AND modal_locations.subscriber = location_histogram.subscriber
+	AND location_histogram.y >= {self.home_this_month}
+), """
+
+        if self.ref_location:
+            sql += f"""
+reference_locations AS (
+    SELECT subscriber, pcod
+    FROM ({self.ref_location.get_query()}) AS tbl
+), last_month_known_homes AS (
+    SELECT location_histogram.subscriber, location_histogram.pcod
+    FROM location_histogram 
+    INNER JOIN  reference_locations USING (subscriber) 
+    INNER JOIN modal_locations USING (subscriber)
+    WHERE modal_locations.pcod = reference_locations.pcod
+    AND location_histogram.y >= {self.home_last_month}
+    AND location_histogram.subscriber NOT IN (SELECT subscriber FROM this_month_known_homes)
+), """
+
+        if self.ref_location is None:
+            sql += f"""
+known_homes AS (
+    SELECT subscriber, pcod
+    FROM this_month_known_homes
+), """
+        else:
+            sql += f"""
+known_homes AS (
+    SELECT subscriber, pcod
+    FROM this_month_known_homes
+    UNION ALL
+    SELECT subscriber, pcod
+    FROM last_month_known_homes
+), """
+
+        sql += f"""
+unknown_homes AS(
+	SELECT DISTINCT(subscriber), 'unknown' AS pcod
```

Comment:
Following on from my earlier comment about using `self.spatial_unit.location_id_columns` instead of `pcod` to handle spatial units other than admin, I've just realised setting the 'unknown' value could be problematic if the spatial unit doesn't have a single text field as the location ID (e.g. VersionedCell, which has two columns, or LonLat, which has a pair of coordinates). Not sure how to cleanly handle that situation.